lasagne/layers/pool.py

```diff
@@ -755,3 +756,161 @@ def get_output_shape_for(self, input_shape):
 
     def get_output_for(self, input, **kwargs):
         return self.pool_function(input.flatten(3), axis=2)
+
+
+def pooling_fixed_output_size(inputs, output_size, pooling_op=T.max):
+    """
+    Performs a pooling operation that results in a fixed output size.
+    Used by SpatialPyramidPoolingLayer. Refer to appendix A in [1]
+
+    Parameters
+    ----------
+    inputs : a tensor with 4 dimensions (N x C x H x W)
+    output_size: integer
+        The output size of the pooling operation
+    pooling_op: a theano op
+        The pooling operation (default: max pooling)
+
+    Returns a tensor with 3 dimensions (N x C x output_size*output_size)
+
+    References
+    ----------
+    .. [1] He, Kaiming et al (2015):
+           Spatial Pyramid Pooling in Deep Convolutional Networks
+           for Visual Recognition.
+           http://arxiv.org/pdf/1406.4729.pdf.
+    """
+    h, w = inputs.shape[2:]
+
+    result = []
+    n = float(output_size)
+
+    for row in range(output_size):
+        for col in range(output_size):
+            start_h = T.floor(row / n * h).astype('int32')
+            end_h = T.ceil((row + 1) / n * h).astype('int32')
+            start_w = T.floor(col / n * w).astype('int32')
+            end_w = T.ceil((col + 1) / n * w).astype('int32')
+
+            pooling_region = inputs[:, :, start_h:end_h, start_w:end_w]
+            this_result = pooling_op(pooling_region, axis=(2, 3))
+            result.append(this_result.dimshuffle(0, 1, 'x'))
+    return T.concatenate(result, axis=2)
```

Comment:
This will result in cascaded concatenation in the pyramid pooling graph. I'm not sure if this is optimized away by Theano. If not, it might be faster to just return the list and have a single concatenation in SpatialPyramidPoolingLayer.